src/aws.ts

```diff
@@ -504,6 +507,237 @@ export namespace $AWS {
+                      typeof output.value !== "object" ||
+                      !output.value
+                    ) {
+                      throw new SynthError(
+                        ErrorCodes.Unexpected_Error,
+                        "Expected an object literal as the first parameter."
+                      );
+                    }
+
+                    return context.stateWithHeapOutput({
+                      Type: "Task",
+                      Resource: `arn:aws:states:::aws-sdk:${sdkIntegrationServiceName}:${methodName}`,
+                      Parameters: output.value,
+                      Next: ASLGraph.DeferNext,
+                    });
+                  });
+                },
+              });
+            },
+          }
+        );
+      },
+    }
+  );
+}
+
+type AWSServiceClass = { new (): AWSService };
+
+/**
+ * First we have to extract the names of all Services in the v2 AWS namespace
+ *
+ * @returns "AccessAnalyzer" | "Account" | ... | "XRay"
+ */
+type ServiceKeys = {
+  [K in keyof typeof AWS]: typeof AWS[K] extends AWSServiceClass ? K : never;
+}[keyof typeof AWS];
+
+type StepFunctionSDK = {
+  [serviceName in ServiceKeys]: typeof AWS[serviceName] extends new (
+    ...args: any[]
+  ) => infer Client
+    ? {
+        [methodName in keyof Client]: StepFunctionMethod<Client[methodName]>;
+      }
+    : never;
+};
+
+interface IamOptions {
+  /**
+   * The resources for the IAM statement that will be added to the state machine
+   * role's policy to allow the state machine to make the API call.
+   *
+   * Use `["*"]` to grant access to all resources (discouraged).
+   *
+   * @example ["arn:aws:s3:::my_bucket"]
+   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_resource.html
+   */
+  iamResources: string[];
+
+  /**
+   * The action for the IAM statement that will be added to the state machine role's
+   * policy to allow the state machine to make the API call.
+   *
+   * By default the action is inferred from the API call (e.g. `$AWS.SDK.CloudWatch.describeAlarms({})` results in `cloudwatch:DescribeAlarms`)
+   *
+   * Use in the case where the IAM action name does not match with the API service/action name
+   * e.g. `s3:ListBuckets` requires `s3:ListAllMyBuckets`.
+   *
+   * @default service:method
+   * @example s3:ListAllMyBuckets
+   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_action.html
+   */
+  iamActions?: string[];
+
+  /**
+   * The iam conditions to apply to the IAM Statement that will be added to the state machine role's
+   * policy to allow the state machine to make the API call.
+   *
+   * By default no conditions are applied.
+   *
+   * @example
+   * ```
+   * {
+   *   "StringEquals" : { "aws:username" : "johndoe" }
+   * }
+   * ```
+   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition.html
+   */
+  iamConditions: Record<string, any>;
```

Comment:
nit: is the `iam` prefix required if the object is already called `Iam*`?

Reply:
I renamed the parameter to `options` and the `IamOptions` type to `SdkCallOptions`. I wanted this second parameter to be the place where all additional stuff (e.g. your idea of the `serviceNameOverride`) would go. That is why I would leave the `iam*` prefix in the new `SdkCallOptions` type